pkg/dryrun/dryrun.go

```diff
@@ -0,0 +1,810 @@
+	}
+
+	mockCtrl := gomock.NewController(nil)
+	mockRecorder := record.NewFakeRecorder(100)
+	d.mockControllerMgr = mock.NewMockManager(mockCtrl)
+	mockCache := mock.NewMockCache(mockCtrl)
+	mockCache.EXPECT().
+		IndexField(gomock.Any(), gomock.Any(), "nodeName", gomock.Any()).
+		Return(nil).
+		AnyTimes()
+	d.mockControllerMgr.EXPECT().GetClient().Return(d.mockClient).AnyTimes()
+	d.mockControllerMgr.EXPECT().GetScheme().Return(scheme.Scheme).AnyTimes()
+	d.mockControllerMgr.EXPECT().GetEventRecorderFor(gomock.Any()).Return(mockRecorder).AnyTimes()
+	d.mockControllerMgr.EXPECT().GetConfig().Return(&rest.Config{
+		Host:    "127.0.0.1",
+		APIPath: "fake",
+	}).AnyTimes()
+	d.mockControllerMgr.EXPECT().SetFields(gomock.Any()).Return(nil).AnyTimes()
+	d.mockControllerMgr.EXPECT().GetCache().Return(mockCache).AnyTimes()
+	d.mockControllerMgr.EXPECT().Add(gomock.Any()).Return(nil).AnyTimes()
+	d.mockControllerMgr.EXPECT().GetLogger().Return(log.Log.WithName("test")).AnyTimes()
+
+	d.mockController = &storagecluster.Controller{Driver: d.mockDriver}
+	d.mockController.SetEventRecorder(mockRecorder)
+	err = d.mockController.Init(d.mockControllerMgr)
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to initialize controller")
+	}
+	d.mockController.SetKubernetesClient(d.mockClient)
+
+	if outputFolder != "" {
+		d.outputFolder = outputFolder
+	}
+
+	d.cluster, err = d.getStorageCluster(storageClusterFile)
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to get storage cluster.")
+	}
+	// Mock migration approved.
+	d.cluster.Annotations[constants.AnnotationMigrationApproved] = "true"
+	d.cluster.Annotations[constants.AnnotationPauseComponentMigration] = "false"
+	d.cluster.Status.Phase = constants.PhaseMigrationInProgress
+	// Disable storage due to the driver will talk to portworx SDK to get storage node information.
+	// However dryrun is before installation so there won't be useful information.
+	// It will also disable PDB.
+	// We will generate portworx pod spec by calling the API directly.
+	d.cluster.Annotations[constants.AnnotationDisableStorage] = "true"
+
+	return nil
+}
+
+// Execute dry run for portworx installation
+func (d *DryRun) Execute() error {
+	err := d.mockClient.Create(context.TODO(), d.cluster)
+	if err != nil {
+		return err
+	}
+
+	if err = d.simulateK8sNode(); err != nil {
+		return err
+	}
+
+	if err = d.installAllComponents(); err != nil {
+		return err
+	}
+
+	objs, err := d.getAllObjects()
+	if err != nil {
+		logrus.WithError(err).Error("failed to get all objects")
+	}
+
+	if err = d.writeFiles(objs); err != nil {
+		return err
+	}
+	logrus.Infof("Operator will deploy %d objects, saved to folder %s", len(objs), d.outputFolder)
+
+	if d.realClient != nil {
+		controller := &storagecluster.Controller{}
+		controller.SetKubernetesClient(d.realClient)
+		h := migration.New(controller)
+		dsObjs, err := h.GetAllDaemonSetObjects(d.cluster.Namespace)
```

Comment:
Are you assuming we only have portworx daemonset in the namespace. Is it possible you have multiple DaemonSets ?

Reply:
yes, there is a ticket to support two daemonset installs, it's not supported yet. we will document this prior to release